fs-api/src/System/FS/API.hs

```diff
@@ -175,3 +181,92 @@ hClose' HasFS { hClose, hIsOpen } h = do
 -- hides an existential @h@ parameter of a 'HasFS'.
 data SomeHasFS m where
   SomeHasFS :: Eq h => HasFS m h -> SomeHasFS m
+
+{-------------------------------------------------------------------------------
+  HasBufFS
+-------------------------------------------------------------------------------}
+
+-- | Abstract interface for performing I\/O using user-supplied buffers.
+--
+-- Note: this interface is likely going to become part of the 'HasFS' interface,
+-- but is separated for now so downstream code does not break because of adding
+-- an additional type parameter.
+--
+-- TODO: add versions that use the file offset that is stored in the file
+-- handle, akin to 'hGetSome' and 'hPutSome'.
+data HasBufFS m h ptr = HasBufFS {
```

Comment:
One thing I'm yet unsure of is whether we need this `ptr` argument. The reason it's currently added is because the pure model can't really be made to handle `Foreign.Ptr` very well. I tried doing that, but it got really messy really fast. I've thought about whether `simHasFS` can handle `Foreign.Ptr`, and I think it can, because we can manipulate `Foreign.Ptr` (though slightly hacky) in `ST` and therefore also in `IOSim`. However, then we still need to provide an abstraction for allocation functions in an arbitrary monad, and so we are adding an abstraction anyway though we got rid of the `ptr` argument... so in any case, it seems we need some abstraction for pointers and/or memory allocation

In any case, this is something to think about, since I'm not yet really happy with the `ptr` argument

Reply:
See https://github.com/input-output-hk/fs-sim/pull/46#issuecomment-1986056449